library/core/src/iter/adapters/intersperse.rs

```diff
@@ -40,37 +40,143 @@ where
         }
     }
 
-    fn fold<B, F>(mut self, init: B, mut f: F) -> B
+    fn fold<B, F>(self, init: B, f: F) -> B
     where
         Self: Sized,
         F: FnMut(B, Self::Item) -> B,
     {
-        let mut accum = init;
+        let separator = self.separator;
+        intersperse_fold(self.iter, init, f, move || separator.clone(), self.needs_sep)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        intersperse_size_hint(&self.iter, self.needs_sep)
+    }
+}
+
+/// An iterator adapter that places a separator between all elements.
+#[unstable(feature = "iter_intersperse", reason = "recently added", issue = "79524")]
+pub struct IntersperseWith<I, G>
```

Comment:
It'd be good to link to `Iterator::intersperse_with()` as the way of creating this struct. (Just like e.g. [`Zip`s documentation](https://doc.rust-lang.org/stable/std/iter/struct.Zip.html) does.). (I noticed `Intersperse` is also missing a link to `Iterator::intersperse()`.)